src/main/scala/com/example/solution/services/Reservations.scala

```diff
@@ -0,0 +1,134 @@
+                                                                  screenings: Screenings[F]
+                                                                ) extends Reservations[F] {
+
+  import ReservationQueries._
+
+  private def areSeatsValid(picked: List[PickedSeat],
+                            available: List[AvailableSeat],
+                            seatsPerRow: Int): Boolean = {
+    val isSeatAlreadyTaken = picked.exists(ps => !available.contains(AvailableSeat(ps.row, ps.seatInRow)))
+    val leftSeats = available diff picked.map(ps => AvailableSeat(ps.row, ps.seatInRow))
+    val leftOneEmptySeat = leftSeats.groupBy(_.row).exists(tuple => {
+      val list = tuple._2
+      list.exists(seat => {
+        val seatInRow = seat.seatInRow
+        (seatInRow > 1 && seatInRow < seatsPerRow) &&
+          (!list.exists(as => as.seatInRow == seatInRow - 1) && !list.exists(as => as.seatInRow == seatInRow + 1))
+      })
+    })
+
+    !(isSeatAlreadyTaken || leftOneEmptySeat)
+  }
+
+
+  private def getSeats(seats: List[PickedSeat]): OptionT[F, NonEmptyList[PickedSeat]] =
+    OptionT(Sync[F].delay(seats.headOption match {
+      case Some(value) => Option(NonEmptyList(value, seats.tail))
+      case None => Option.empty[NonEmptyList[PickedSeat]]
+    }))
+
+
+  override def makeReservation(data: ReservationData): F[Option[MadeReservation]] = {
+
+    val userWithSeats = (for {
+      user <- OptionT(Sync[F].delay(User.mkUser(UserName(data.name), UserSurname(data.surname))))
+      rd <- OptionT(screenings.pickScreening(data.screeningId))
+      if rd.screeningTime.minus(15L, ChronoUnit.MINUTES).compareTo(LocalDateTime.now()) > 0
+
+      pickedSeats <- getSeats(data.pickedSeats)
+      seatsWithTickets <- OptionT(Sync[F].delay(pickedSeats.map(ps => for {
+        seat <- Option(ps)
+        ticket <- Ticket.fromString(ps.ticketType)
+      } yield (seat, ticket)).sequence)) if areSeatsValid(seatsWithTickets.map(_._1).toList, rd.availableSeats, rd.room.seatsPerRow)
+
+    } yield (user, seatsWithTickets)).value
+
+    userWithSeats.flatMap {
+      case None => Sync[F].delay(Option.empty[MadeReservation])
+      case Some(tuple) =>
+        val (user, seatsWithTickets) = (tuple._1, tuple._2.toList)
+        sessionPool.use { session =>
+          session.prepare(createReservation).use { cmd =>
+            val totalCost = seatsWithTickets.map(_._2.price).sum
+            cmd.execute((user, data.screeningId), totalCost).void
+          } *> session.execute(selectMaxId).map(_.head).flatMap(id => {
+            seatsWithTickets.traverse(tuple => {
+              val (ps, ticket) = tuple
+              session.prepare(createTakenSeat).use { cmd =>
+                cmd.execute((id, ps), ticket).void
+              }
+            })
+          }).map(_ => MadeReservation(
+            seatsWithTickets.map(_._2.price).sum,
+            LocalDateTime.now().plus(1L, ChronoUnit.HOURS))
```

Comment:
`now()` method can be problematic. Do you know why?

Reply:
I guess I should have used Instant instead, I believe the time returned by LocalDateTime.now() might be dependent on OS settings, but I'm not sure about that.